README.md
# Rpkg: command-line interface to the R package manager

## What is this?

This is a very simple command-line interface to the R package manager. 
Typically this system is accessed using functions in the R console like 
`install.packages()`. I wrote this tool because sometimes I find it 
inconvenient to open an R console to manage the R package library.

The interface is currently very minimal, and several features have not yet 
been implemented. On the other hand, it has no dependencies other than R 
itself. The vision is twofold:
1. Create a simple, unified interface to the package management features 
   that are built into R.
2. Reach a stable, complete 1.0 interface that will not need to change much. 
   In 10 years, when GitHub has gone the way of SourceForge, you should be 
   able do download and run this program without serious problems, as long as 
   R itself does not undergo radical changes (which it historically has not).

If you like this program, or please feel free to submit pull requests or post 
issues with bugs reports, feature requests, etc. I can't guarantee that I'll 
have much time to work on it, but I do use it myself on a regular basis. So if 
there is a substantial improvement to be made or a serious bug to fix, I 
will be motivated to integrate the changes. See `TODO.md` for some of the 
roadmap, as well as the various `TODO`s in the source code itself.

### Show-off example

```
$ Rpkg install forecast

Installing package into ‘/usr/local/lib/R/3.4/site-library’
(as ‘lib’ is unspecified)
trying URL 'https://cloud.r-project.org/src/contrib/forecast_8.1.tar.gz'
Content type 'application/x-gzip' length 783567 bytes (765 KB)
==================================================
downloaded 765 KB

* installing *source* package ‘forecast’ ...
** package ‘forecast’ successfully unpacked and MD5 sums checked
** libs
** R
** data
*** moving datasets to lazyload DB
** inst
** byte-compile and prepare package for lazy loading
** help
*** installing help indices
*** copying figures
** building package indices
** installing vignettes
** testing if installed package can be loaded
* DONE (forecast)

The downloaded source packages are in
	‘/private/var/folders/4j/n8nxnsy12y92t475g318yptm0000gn/T/RtmpKRM4Ne/downloaded_packages’
```


## Installation instructions

1. `git clone` this repo or download and extract a tagged release.
2. `make`
3. `make install`, using `sudo` if necessary on your system. By default this installs to `/usr/local/bin/Rpkg`, but you can change that location with the `RPKG_INSTALL_PREFIX` environment variable.

I've personally tested the Makefile on macOS GNU Make 3.81 and vanilla GNU Make 4.2.1. It's very straightforward, but your mileage might vary with BSD Make or other/ancient Makes.

I've also only personally tested this on R version 3.4.1. I don't *think* I'm using anything specific to recent R versions, but if anything breaks in old versions, please file an issue.

In the code and documentation, I assume that you have copied or linked 
`rpkg.R` with the name `Rpkg`, and that it is user-executable. Any other usage 
is untested and unsupported.

This script is intended to be run by `Rscript --no-save`. Note that `Rscript`
already implies `--slave` and `--no-restore`. It _does_ source your init file,
i.e. `Rprofile`. This is to ensure that any user configuration like
`options(repos = c(CRAN = "https://cloud.r-project.org"))` is respected.
HOWEVER, Rpkg is **not** robust to excessive tinkering, and if something breaks
try `Rscript --no-save --no-init rpkg.R` before reporting a bug.

The default CRAN repository when using this tool is
`"https://cloud.r-project.org"`. The option to use a non-CRAN repository will
become available after proper command-line parsing has been implemented.


## Usage instructions

Using Rpkg is much like any other package manager: `yum`, `brew`, `pip`, 
`npm`, etc. Commands take the form:

```shell
Rpkg <main_options> <subcommand> <subcommand_options> <arguments>
```

The `main_options` are Unix-style flags (even on Windows, they are currently 
hardcoded to use dashes and not slashes) that change the behavior of the main 
`Rpkg` command. The `subcommand_options` apply specifically to the subcommand 
being called.

The `Rpkg help` subcommand prints a list of available commands and options for 
the `Rpkg` script itself. In the future, `Rpkg help <subcommand>` will print 
similar information for the listed subcommand.

As a reference, the `Rpkg help` string from the source code is copied below:

```
Commands:
    help                      [ subcommand ]
      NOTE: not fully implemented
    ( install | add )         pkg_name ...
    ( uninstall | remove )    pkg_name ...
    ( update | upgrade )      [ --all ] pkg_name ...
    reinstall                 pkg_name ...
    outdated
    list
    info                      pkg_name ...
    search                    ( string | /regex/ )
      NOTE: regex is PCRE, case-sensitive only if uppercase character is detected

Options:
    -V / --version
    -h / --help
```


### Examples

Install `caret` and `shiny` from CRAN:

```shell
Rpkg install caret shiny
Rpkg add caret shiny
```

Update `caret` and `shiny` to latest version on CRAN:

```shell
Rpkg update caret shiny
Rpkg upgrade caret shiny
```

Update all packages:

```shell
Rpkg upgrade --all
Rpkg update --all
```

Uninstall `caret` and `shiny`:

```shell
Rpkg uninstall caret shiny
```

Check for outdated packages, get info about the `forestFloor` package, list 
installed packages:

```shell
Rpkg outdated
Rpkg info forestFloor
Rpkg list
```

Search CRAN:

```shell
Rpkg search ggplot
Rpkg search '/^z[^aeiou]/' # pattern is a regular expression if wrapped in slashes
Rpkg search '/^GG/'        # case-sensitive only if the pattern has an upper case character
```

## Contributing

Pull requests welcome! Development happens on the `dev` branch, which is merged to `master` for tagged releases.

<!-- vim: ft=pandoc:
-->


NEWS.md
TODO, like everything else :)


package/Rpkg/R/rpkg.R
# TODO:
#   - move all side effects (i.e. "catn(...)") to CLI
#   - return S3 (?) classes/structures in this code
#   - define format() and/or print() methods for S3 classes

VERSION <- "0.5.0-rc01"


## TODO: use optparse or getopt library and .libPaths() so user doesn't have to have pkgs installed
## e.g.:
##     .libPaths('/usr/local/lib/rpkg/library')
## and then can call getopt::getopt() or optparse::OptionParser as usual
## NOTE: be careful with this because it could really swell installation size


#### Helpers ----


# man 3 sysexits
sysexits <- new.env()
sysexits$EX_OK          <- 0L
sysexits$EX_USAGE       <- 64
#sysexits$EX_DATAERR     <- 65
sysexits$EX_NOINPUT     <- 66
#sysexits$EX_NOUSER      <- 67
#sysexits$EX_NOHOST      <- 68
sysexits$EX_UNAVAILABLE <- 69
sysexits$EX_SOFTWARE    <- 70
#sysexits$EX_OSERR       <- 71
#sysexits$EX_OSFILE      <- 72
#sysexits$EX_CANTCREAT   <- 73
#sysexits$EX_IOERR       <- 74
#sysexits$EX_TEMPFAIL    <- 75
#sysexits$EX_PROTOCOL    <- 76
#sysexits$EX_NOPERM      <- 77
#sysexits$EX_CONFIG      <- 78


unattr <- function(x) {
  `attributes<-`(x, NULL)
}


`%!in%` <- function(x, table) {
  match(x, table, nomatch = 0L) == 0L
}


catn <- function(...) {
  cat(..., "\n")
}


cat0 <- function(...) {
  cat(..., sep = "")
}


cat0n <- function(...) {
  cat0(..., "\n")
}


exit <- function(status = 0L, msg = NULL, con = if (length(status) && is.numeric(status) && status) stderr() else stdout()) {
  if (!is.null(msg) && length(msg) > 0) {
    cat0n(unlist(msg), file = con)
  }
  quit(save = "no", status = status, runLast = FALSE)
}


text_search <- function(x, pattern, perl = TRUE, ignore.case = TRUE) {
  if (length(pattern) != 1 || !is.character(pattern)) {
    stop("Invalid pattern")
  }

  regex_match <- regexpr("^/(.*)/$", pattern, perl = TRUE)
  regex_start <- attr(regex_match, "capture.start")
  if (regex_start < 0) {
    # is not regex
    fixed <- TRUE
    perl <- FALSE
    ignore.case <- FALSE
    do_lowercase <- ignore.case
  } else {
    # is regex
    pattern <- substr(pattern, regex_start, regex_start + attr(regex_match, "capture.length") - 1)
    fixed <- FALSE
    perl <- TRUE
    do_lowercase <- FALSE
    ignore.case <- !grepl("[[:upper:]]", pattern)
  }

  ix <- grep(if (do_lowercase) tolower(pattern) else pattern, if (do_lowercase) tolower(x) else x,
             perl = perl, fixed = fixed, ignore.case = ignore.case)
  x[ix]
}


#' Get names of installed packags
#'
#' @param ... Arguments passed to \code{\link[utils]{installed.packages()}}
#' @return Character vector of installed package names
get_installed_package_names <- function(...) {
  rownames(utils::installed.packages(...))
}


#' Validate which packages are not currently installed
#'
#' @param packages Character vector of package names
#' @return Character vector of package names that are not yet installed
#' @note Side effect: prints a message and names of currently installed packages to Stderr
validate_notyet_installed <- function(packages) {
  all_installed <- get_installed_package_names()
  already_installed <- intersect(packages, all_installed)

  if (length(already_installed)) {
    message("These packages are already installed and will be skipped:")
    cat(already_installed, file = stderr(), sep = "\n")
    not_yet_installed <- setdiff(packages, already_installed)
  } else {
    not_yet_installed <- packages
  }

  not_yet_installed
}


#' Validate which packages are currently installed
#'
#' @param packages Character vector of package names
#' @return Character vector of package names that are already installed
#' @note Side effect: prints a message and names of not-yet installed packages to Stderr
validate_already_installed <- function(packages) {
  all_installed <- get_installed_package_names()
  not_yet_installed <- setdiff(packages, all_installed)

  if (length(not_yet_installed)) {
    message("These packages are not installed and will be skipped:")
    cat(not_yet_installed, file = stderr(), sep = "\n")
    already_installed <- intersect(packages, all_installed)
  } else {
    already_installed <- packages
  }

  already_installed
}


#### R package management ----


#' Install packages
#'
#' @param packages Character vector of package names
#' @param opts List of named options, passed to \code{\link[utils]{install.packages()}}
#' @return Result of \code{\link[utils]{install.packages}}
pkg_install <- function(packages, opts = list()) {
  packages_to_install <- validate_notyet_installed(packages)

  if (!length(packages_to_install)) {
    exit(sysexits$EX_NOINPUT, "No packages specified")
  }

  do.call(utils::install.packages, c(list(packages_to_install), opts))
}


#' Remove packages
#'
#' @param packages Character vector of package names
#' @param opts List of named options, passed to \code{\link[utils]{remove.packages()}}
#' @return Result of \code{\link[utils]{remove.packages()}}
pkg_remove <- function(packages, opts = list()) {
  packages_to_remove <- validate_already_installed(packages)

  if (!length(packages_to_remove)) {
    exit(sysexits$EX_NOINPUT, "No packages specified")
  }

  do.call(utils::remove.packages, c(list(packages_to_remove), opts))
}


#' Update packages
#'
#' @param packages Character vector of package names
#' @param opts List of named options, passed to \code{\link[utils]{update.packages()}}
#' @return Result of \code{\link[utils]{update.packages()}}
# TODO: warn if --all passed along with args
pkg_update <- function(packages, opts = list()) {
  if (length(packages) == 0 && "all" %in% names(opts) && opts$all) {
    catn("Updating all packages")
    opts$oldPkgs <- NULL
  } else {
    packages_to_update <- validate_already_installed(packages)

    if (!length(packages_to_update)) {
      exit(sysexits$EX_NOINPUT, "No packages specified")
    }

    opts$oldPkgs <- packages_to_update
  }

  opts$all <- NULL
  opts$ask <- FALSE
  do.call(utils::update.packages, opts)
}


#' Check for outdated packages
#'
#' @param packages Character vector of package names
#' @param opts List of named options, passed to \code{\link{old.packages()}}
#' @return Result of \code{\link[utils]{old.packages()}}
pkg_outdated <- function(packages, opts = list()) {
  if (length(packages)) {
    exit(sysexits$EX_UNAVAILABLE, "Package selection not implemented for this subcommand")
  }

  out <- do.call(utils::old.packages, opts)
  if (is.null(out)) {
    catn("All packages are up-to date")
  } else {
    out <- out[, c("Installed", "ReposVer", "Repository")]
    colnames(out) <- c("LocalVersion", "RemoteVersion", "Repo")
    rownames(out) <- rep("", nrow(out))
    print(out, quote = FALSE)
  }
}


pkg_list <- function(packages, opts = list()) {
  cat(get_installed_package_names(), sep = "\n")
}


# TODO: fuzzy search with --fuzzy, ignore case with --ignore-case, regex search with --regex (instead of detecting '/pattern/')
pkg_info <- function(packages, opts = list()) {
  if (length(packages) < 0) {
    exit(sysexits$EX_NOINPUT, "No packages specified")
  }

  pkg_avail <- as.data.frame(available.packages())
  pkg_inst <- as.data.frame(installed.packages())
  pkgs <- merge(pkg_avail, pkg_inst, by = "Package", suffixes = c("", "_DELETEME"), all = TRUE)
  pkgs <- pkgs[, grep("_DELETEME", names(pkgs), invert = TRUE)]

  # TODO: parse and compare installed vs latest available in repo (e.g. version numbers)
  for (package in packages) {
    if (package %!in% pkgs$Package) {
      catn(sprintf("Unknown package: %s. Try: Rpkg search %s", package, package))
    } else {
      catn(sprintf("~*~*~ %s ~*~*~", package))
      cat(sprintf("%s: %s", names(pkgs), as.character(pkgs[pkgs$Package == package,,drop = TRUE])), sep = "\n")
      cat("\n")
    }
  }
}


# Convenience alias for "uninstall" followed by "install"
pkg_reinstall <- function(packages, opts = list()) {
  for (package in packages) {
    pkg_remove(package)
    pkg_install(package, opts)
  }
}


pkg_search <- function(patterns, opts = list()) {
  if (length(patterns) < 0) {
    exit(sysexits$EX_NOINPUT, "No patterns specified")
  }

  avail <- available.packages()
  pkgs <- character(0)
  for (pattern in patterns) {
    pkgs <- c(pkgs, text_search(avail[,"Package"], pattern))
  }
  cat(sort(unique(pkgs)), sep = "\n")
}


Rpkg.R
#! Rscript --no-save --no-restore

# TODO: main() CLI argument handling
# use optparse() library?


..VERSION.. <- 0.2


exit <- function(status = 0L, msg = NULL, con = if (status) stderr() else stdout()) {
  if (!is.null(msg)) {
    cat(msg, "\n", file = con)
  }

  quit(save = "no", status = status, runLast = FALSE)
}


#' Get names of installed packags
#'
#' @param ... Arguments passed to \code{\link{installed.packages()}}
get_installed_packages <- function(...) {
  rownames(utils::installed.packages(...))
}


#' Install packages
#'
#' @param packages Character vector of package names
#' @param opts List of named options, passed to \code{\link{install.packages()}}
pkg_install <- function(packages, opts = list()) {
  installed <- get_installed_packages()
  already_installed <- intersect(packages, installed)

  if (length(already_installed)) {
    message("These packages already been installed and will be skipped:")
    cat(already_installed, "\n", file = stderr())
    packages_to_install <- setdiff(packages, already_installed)
  } else {
    packages_to_install <- packages
  }

  if (!length(packages_to_install)) {
    exit(66, "No packages specified")
  }
  do.call(utils::install.packages, c(list(packages_to_install), opts))
}


#' Remove packages
#'
#' @param packages Character vector of package names
#' @param opts List of named options, passed to \code{\link{remove.packages()}}
pkg_remove <- function(packages, opts = list()) {
  installed <- get_installed_packages()
  notyet_installed <- setdiff(packages, installed)

  if (length(notyet_installed)) {
    message("These packages are not installed and will be skipped:")
    cat(notyet_installed, "\n", file = stderr())
    packages_to_remove <- intersect(packages, installed)
  } else {
    packages_to_remove <- packages
  }

  if (!length(packages_to_remove)) {
    exit(66, "No packages specified")
  }
  do.call(utils::remove.packages, c(list(packages_to_remove), opts))
}


#' Update packages
#'
#' @param packages Character vector of package names
#' @param opts List of named options, passed to \code{\link{update.packages()}}
pkg_update <- function(packages, opts = list()) {
  installed <- get_installed_packages()
  notyet_installed <- setdiff(packages, installed)

  if (length(notyet_installed)) {
    message("These packages are not installed and will be skipped:")
    cat(notyet_installed, "\n", file = stderr())
    packages_to_update <- intersect(packages, installed)
  } else {
    packages_to_update <- packages
  }

  if (!length(packages_to_update)) {
    exit(66, "No packages specified")
  }
  opts$ask <- FALSE
  opts$oldPkgs <- NULL
  do.call(utils::update.packages, c(list(oldPkgs = packages_to_update), opts))
}


#' Check for outdated packages
#'
#' @param packages Character vector of package names
#' @param opts List of named options, passed to \code{\link{old.packages()}}
pkg_outdated <- function(packages, opts = list()) {
  if (length(packages)) {
    opts$instPkgs <- packages
  }

  do.call(utils::old.packages, opts)
}

pkg_search <- function(packages, opts = list()) {
  exit(69, "Package search not implemented")
}


rpkg_version <- ..VERSION..
rpkg_help <- sprintf(
"Rpkg version %s

Commands:
    help
    install / add
    upgrade / update
    uninstall / remove
    search (not implemented)

Options:
    -V / --version
    -h / --help",
rpkg_version)


main <- function() {
  cli_args <- commandArgs(trailingOnly = TRUE)

  cmd <- cli_args[1]
  pkgs <- cli_args[-1]

  if (any(cli_args %in% c("-V", "--version"))) {
    exit(0, rpkg_version)
  }

  if (any(cli_args %in% c("-h", "--help"))) {
    exit(0, rpkg_help)
  }

  opts <- list()

  # TODO: handle "--" to stop processing option flags
  # TODO: handle "ask = TRUE" (why doesn't it work, anyway?)
  # TODO: handle verbosity
  # TODO: self-update?
  switch(cmd,
    "install"   = pkg_install(pkgs, opts),
    "add"       = pkg_install(pkgs, opts),

    # TODO: handle "--all"
    "update"    = pkg_update(pkgs, opts),
    "upgrade"   = pkg_update(pkgs, opts),

    # TODO: handle pkg names
    "outdated"  = pkg_outdated(NULL, opts),

    "remove"    = pkg_remove(pkgs, opts),
    "uninstall" = pkg_remove(pkgs, opts),

    "search"    = pkg_search(pkgs, opts),

    "help"      = exit(0, rpkg_help),
    `NA`        = exit(64, "No command specified"),
                  exit(64, sprintf("Invalid command: %s", cmd))
  )
}

main()
